### ISSUE TITLE
Bug: `CheckoutLineDiscount` is not deleted when sale is not applicable on checkout's channel anymore

### ISSUE THREAD

**Issue by IKarbowiak** [saleor/saleor#14002]
### What are you trying to achieve?

`CheckoutLineDiscount` are not removed when the sale is not valid in checkout's channel anymore.

### Steps to reproduce the problem

1. Create a `Sale`
2. Set channel availability
3. Assign some products to this sale
4. Create checkout with products that is included in this sale
5. Before completing checkout drop sale channel listing for the checkout's channel
6. Force checkout price recalculation and check the discounts. The checkout line discount will be still visible

### What did you expect to happen?

The `CheckoutLineDiscount` should be deleted when it's not valid anymore.

[create_or_update_discount_objects_from_sale_for_checkout](https://github.com/saleor/saleor/blob/main/saleor/discount/utils.py#L622) is responsible for creating and updating those instances, it should removed them as well.

Consider forcing related checkout price recalculation during sale channel listings deletion

### Environment

Saleor version: main

**Comment by korycins** [comment]
Tested on 3.14, result with the same issue

**Comment by simon-online** [comment]
This still seems to be an issue in core v3.19.18.

When I delete a discount there is no worker task event logged so all the old discounted product pricing remains.

**Comment by maarcingebala** [comment]
@saleor/shopex-py Can you check this one?

**Comment by vhejda** [comment]
There seems to be a bigger issue than affecting checkouts. I just found out that product variant pricing is still showing a discount that has been long de-assigned. Even after deleting the sale completely, the products are still showing the discounted price, coming in from `ProductVariant.pricing.price`.

Saleor version 3.20.0

**Comment by IKarbowiak** [comment]
Hi @vhejda,
Please ensure that your celery worker is running, as the price recalculation is done in the background by the celery worker. there are two tasks responsible for that: `update_variant_relations_for_active_promotion_rules_task` and `recalculate_discounted_price_for_products_task`

**Comment by vhejda** [comment]
> Hi @vhejda, Please ensure that your celery worker is running, as the price recalculation is done in the background by the celery worker. there are two tasks responsible for that: `update_variant_relations_for_active_promotion_rules_task` and `recalculate_discounted_price_for_products_task`

Hi @IKarbowiak , thanks for the reply. We have the worker running, but it doesn't seem to be picking up these tasks. Only time I see `recalculate_discounted_price_for_products_task` is immediately after the worker has started, where it says:

```
INFO celery.beat Scheduler: Sending due task recalculate-discounted-price-for-products
```

So I have now restarted the worker container and this appeared again and recalculated the prices. But otherwise the only way the prices are recalculated without restarting the worker is changing the price on the variant. That removes / adds the discounts.

Do you know what might be the issue? Really appreciate your help.

**Comment by vhejda** [comment]
> > > Hi @vhejda, Please ensure that your celery worker is running, as the price recalculation is done in the background by the celery worker. there are two tasks responsible for that: `update_variant_relations_for_active_promotion_rules_task` and `recalculate_discounted_price_for_products_task`
> > 
> > 
> > Hi @IKarbowiak , thanks for the reply. We have the worker running, but it doesn't seem to be picking up these tasks. Only time I see `recalculate_discounted_price_for_products_task` is immediately after the worker has started, where it says:
> > ```
> > INFO celery.beat Scheduler: Sending due task recalculate-discounted-price-for-products
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > So I have now restarted the worker container and this appeared again and recalculated the prices. But otherwise the only way the prices are recalculated without restarting the worker is changing the price on the variant. That removes / adds the discounts.
> > Do you know what might be the issue? Really appreciate your help.
> 
> This may be the answer you are looking for: [#15892 (comment)](https://github.com/saleor/saleor/issues/15892#issuecomment-2081122018)

Thanks @simon-online , looks like I have issue with the Docker compose setup in general. But I noticed your comment in another issue https://github.com/saleor/saleor/discussions/14617#discussioncomment-9172419 

Sorry if it's a stupid question but I'm not familiar with Django... What is this "beat process" you're referring to? I have 2 containers, one for the api and other for the worker, with command `celery -A saleor --app=saleor.celeryconf:app worker --loglevel=info -B`.

But you're mentioning `celery -A saleor --app=saleor.celeryconf:app beat --loglevel=debug`. From the context I understand this is some periodic scheduler that I am likely missing. Do you create a 3rd container for this? Or what is the "normal" setup?

I am finding it hard to find any documentation about this, only the `saleor-platform` repository, which is not intended for production use.

**Comment by simon-online** [comment]
> > > > Hi @vhejda, Please ensure that your celery worker is running, as the price recalculation is done in the background by the celery worker. there are two tasks responsible for that: `update_variant_relations_for_active_promotion_rules_task` and `recalculate_discounted_price_for_products_task`
> > > 
> > > 
> > > Hi @IKarbowiak , thanks for the reply. We have the worker running, but it doesn't seem to be picking up these tasks. Only time I see `recalculate_discounted_price_for_products_task` is immediately after the worker has started, where it says:
> > > ```
> > > INFO celery.beat Scheduler: Sending due task recalculate-discounted-price-for-products
> > > ```   
> > >   
> > > So I have now restarted the worker container and this appeared again and recalculated the prices. But otherwise the only way the prices are recalculated without restarting the worker is changing the price on the variant. That removes / adds the discounts.
> > > Do you know what might be the issue? Really appreciate your help.
> > 
> > 
> > This may be the answer you are looking for: [#15892 (comment)](https://github.com/saleor/saleor/issues/15892#issuecomment-2081122018)
> 
> Thanks @simon-online , looks like I have issue with the Docker compose setup in general. But I noticed your comment in another issue [#14617 (reply in thread)](https://github.com/saleor/saleor/discussions/14617#discussioncomment-9172419)
> 
> Sorry if it's a stupid question but I'm not familiar with Django... What is this "beat process" you're referring to? I have 2 containers, one for the api and other for the worker, with command `celery -A saleor --app=saleor.celeryconf:app worker --loglevel=info -B`.
> 
> But you're mentioning `celery -A saleor --app=saleor.celeryconf:app beat --loglevel=debug`. From the context I understand this is some periodic scheduler that I am likely missing. Do you create a 3rd container for this? Or what is the "normal" setup?
> 
> I am finding it hard to find any documentation about this, only the `saleor-platform` repository, which is not intended for production use.

The celery beat process is started automatically on the worker so you don't need to set up any more containers.

Just make sure you include the following line in your `settings.py` file.
```CELERY_BEAT_SCHEDULER = "saleor.schedulers.schedulers.PersistentScheduler"```

**Comment by vhejda** [comment]
> > > > > Hi @vhejda, Please ensure that your celery worker is running, as the price recalculation is done in the background by the celery worker. there are two tasks responsible for that: `update_variant_relations_for_active_promotion_rules_task` and `recalculate_discounted_price_for_products_task`
> > > > 
> > > > 
> > > > Hi @IKarbowiak , thanks for the reply. We have the worker running, but it doesn't seem to be picking up these tasks. Only time I see `recalculate_discounted_price_for_products_task` is immediately after the worker has started, where it says:
> > > > ```
> > > > INFO celery.beat Scheduler: Sending due task recalculate-discounted-price-for-products
> > > > ```
> > > > 
> > > > 
> > > >     
> > > >       
> > > >     
> > > > 
> > > >       
> > > >     
> > > > 
> > > >     
> > > >   
> > > > So I have now restarted the worker container and this appeared again and recalculated the prices. But otherwise the only way the prices are recalculated without restarting the worker is changing the price on the variant. That removes / adds the discounts.
> > > > Do you know what might be the issue? Really appreciate your help.
> > > 
> > > 
> > > This may be the answer you are looking for: [#15892 (comment)](https://github.com/saleor/saleor/issues/15892#issuecomment-2081122018)
> > 
> > 
> > Thanks @simon-online , looks like I have issue with the Docker compose setup in general. But I noticed your comment in another issue [#14617 (reply in thread)](https://github.com/saleor/saleor/discussions/14617#discussioncomment-9172419)
> > Sorry if it's a stupid question but I'm not familiar with Django... What is this "beat process" you're referring to? I have 2 containers, one for the api and other for the worker, with command `celery -A saleor --app=saleor.celeryconf:app worker --loglevel=info -B`.
> > But you're mentioning `celery -A saleor --app=saleor.celeryconf:app beat --loglevel=debug`. From the context I understand this is some periodic scheduler that I am likely missing. Do you create a 3rd container for this? Or what is the "normal" setup?
> > I am finding it hard to find any documentation about this, only the `saleor-platform` repository, which is not intended for production use.
> 
> The celery beat process is started automatically on the worker so you don't need to set up any more containers.
> 
> Just make sure you include the following line in your `settings.py` file. `CELERY_BEAT_SCHEDULER = "saleor.schedulers.schedulers.PersistentScheduler"`

Hmm, I don't have a `settings.py` file, only a docker image (`ghcr.io/saleor/saleor:3.20.0`) and `.env` file with configuration. I tried to add the `CELERY_BEAT_SCHEDULER = "saleor.schedulers.schedulers.PersistentScheduler"` into the env vars but it had no effect.
I really had to add a 3rd container with saleor core and the beat command, and now finally I can see in the worker's log that every minute the tasks are triggered, as opposed to only after the worker is started.
I am sure this is not how it's supposed to be setup, but for now I have a working thing...

It's quite frustrating that the documentation doesn't even mention anything like the worker at all: https://docs.saleor.io/setup/docker-images .

Reading it would make you believe that you simply need the saleor core and saleor dashboard containers and you're good to run.
